modal/_runtime/user_code_imports.py

```diff
@@ -189,6 +339,54 @@ def get_finalized_functions(
             finalized_functions[method_name] = finalized_function
         return finalized_functions
 
+    @contextmanager
+    def execution_context(
+        self,
+        event_loop: UserCodeEventLoop,
+        container_io_manager: "modal._runtime.container_io_manager.ContainerIOManager",
+    ) -> Generator[dict[str, "FinalizedFunction"], None, None]:
+        # 1. Pre-snapshot Enter
+        # Identify all "enter" methods that need to run before we snapshot.
+        if not self.function_def.is_auto_snapshot:
+            pre_snapshot_methods = _find_callables_for_obj(
+                self.user_cls_instance, _PartialFunctionFlags.ENTER_PRE_SNAPSHOT
+            )
+            call_lifecycle_functions(event_loop, container_io_manager, list(pre_snapshot_methods.values()))
+
+        # 2. Snapshot
+        # If this container is being used to create a checkpoint, checkpoint the container after
+        # global imports and initialization. Checkpointed containers run from this point onwards.
+        snapshot_callback(container_io_manager, self.function_def)
+
+        # 3. Breakpoint wrapper
+        create_breakpoint_wrapper(container_io_manager)
+
+        # 4. Post-snapshot Enter
+        # Identify the "enter" methods to run after resuming from a snapshot.
+        if not self.function_def.is_auto_snapshot:
+            post_snapshot_methods = _find_callables_for_obj(
+                self.user_cls_instance, _PartialFunctionFlags.ENTER_POST_SNAPSHOT
+            )
+            call_lifecycle_functions(event_loop, container_io_manager, list(post_snapshot_methods.values()))
+
+        # 5. Get Functions & Start Lifespan
+        with container_io_manager.handle_user_exception():
+            finalized_functions = self.get_finalized_functions(self.function_def, container_io_manager)
+
+        def exit_callback():
```

Comment:
there is something about the asymmetry in enter vs exit here that itches (the exit callback being a callback and not just part of the function's own lifecycle context manager.

If possible, it would be very nice to have the flow here defined as something like this, clearly indicating the ordering and scope of every operation:

```py
with lifecycle_presnapshot():
     maybe_snapshot()
     with lifecycle_postsnapshot():
            with lifecycle_asgi():
                    yield
                    disable_signals()  # ahead of all exit handlers
volume_commit()
enable_signals()
```